### ISSUE TITLE
It is not possible to safely use libjpeg-turbo in multithreaded application through the turbo API

### ISSUE THREAD

**Issue by pac-work** [libjpeg-turbo/libjpeg-turbo#396]
Although it is obvious that authors tried to progress in the direction of making it possible to use the library in applications which use multiple threads (b9ab64d8dbee2db829e59357d335c151680f44f0), there are unfortunately still some loose ends. One of those loose ends is global variable `errStr` (https://github.com/libjpeg-turbo/libjpeg-turbo/blob/2.0.3/turbojpeg.c#L58).

There are at least 2 problems which make it totally impossible to safely use the library in MT applications: 

1. Usage of `errStr` bound to the handle is not strict, for example: https://github.com/libjpeg-turbo/libjpeg-turbo/blob/2.0.3/turbojpeg.c#L1911

2. Global handle, which is accessed through the `THROWG` macro (https://github.com/libjpeg-turbo/libjpeg-turbo/blob/2.0.3/turbojpeg.c#L167), is used all over the place, even inside functions which actually do take handle. For example, `tjBufSize` uses `THROWG` (https://github.com/libjpeg-turbo/libjpeg-turbo/blob/2.0.3/turbojpeg.c#L508), but `tjBufSize` itself is used in `tjCompress2`, which would otherwise take handle, but the fact that it uses `tjBufSize`, which is not thread safe, makes `tjCompress2` not thread safe as well. And more like this...

Any 2 threads accessing `errStr` at the same time immediately cause UB (https://en.wikipedia.org/wiki/Undefined_behavior), which may be especially dangerous in today's world of powerful compiler optimizations, especially link time optimizations.

Potential solutions:
1. Completely remove the global `errStr` to prevent any potential thread races. This would of course decrease the amount of information obtainable during error conditions.
2. Allow the library user to supply his own callback function for handling global errors. All the access to the global `errStr` should go through that function. The default function (if user does not supply anything else) would behave as till now.

**Comment by dcommander** [comment]
The TurboJPEG API evolved rather than was designed, so it is not perfect.  However, I disagree that these issues are fundamental impediments to using that API in a multithreaded program. 
 Specific comments:

Use of `THROWG()`:

This only occurs under one of the following circumstances:
- if an invalid argument is passed to a function that doesn't accept a handle (including arguments that would cause the return value to overflow)
- if an invalid handle is passed to a function that accepts a handle
- if an invalid bitmap file is passed to `tjLoadImage()`
- if a failure occurs while initializing the bitmap writer (`tjSaveImage()`)
- if a memory allocation failure occurs (`tjLoadImage()`, `tjInit*()`)

Those are all unrecoverable errors that, for the most part, would only occur if the calling program was buggy.  Solution: fix the calling program.  I don't know of any other way to handle such errors.

Use of `errStr` in `tjTransform()`:

I agree that this is incorrect.  It was apparently an oversight.  I will fix it in 2.0.4.

Given that `errStr` is of fixed length, I posit that the worst case is two threads encountering a fatal error at the same time and writing to the string simultaneously, thus producing a junk string.  We could probably get around this by using thread-local storage, but that would introduce a dependency on libpthread, which is undesirable (and, of course, we would have to use a different TLS mechanism for Windows.)

**Comment by pac-work** [comment]
Hello and thank you for your response. I think that TurboJPEG API is actually quite easy to use and I am even more impressed that it turned out so good without more upfront design as you say. However in my eyes, the multi-threading issue just needs proper solution.

In my experience, attitude of “it kind of works now, with the current toolchains I tested that with, although it is wrong according to the standard” inevitably leads to problems in the future. And I am saying this as a person who back then wasted more than enough hours after we upgraded gcc from 6 to 7 and one of the major wide-spread open-source distributed databases access driver broke and started to exhibit “random” failures. Compiler simply started to be more aggressive with some optimizations and simple naive thinking about “why is breaking the standard not a problem in this case” stopped working, because compiler on the assembler level of the optimized code was doing something which was far far away from any straightforward thinking of any human. And this was not the only case I encountered this in my life.

The fact that thread sanitizer lights up as a Christmas tree during tests (to be relevant to the current time of the year :) when executing code which uses TurboJPEG API is a very simple no-go in any environment where stability is required, such as 24/7 running embedded systems.

I see why it is not a good idea to have libpthread dependency and agree with that. I just think that simply having an API which causes undefined behavior is pretty dangerous. Your post mentions that this happens only in cases which are otherwise problematic anyway, but the fact that not all of these cases are actually within control of calling application (not possible to write to a file, for example) is a bit alarming, even security-wise.

I was thinking about how to solve the problem with minimum API changes while preserving the current functionality (because not all programs are multi-threaded) and it seems to me that simply allowing the library user to set his own “global error reporting” function would be enough. There should be no direct  `snprintf` into `errStr` anywhere, that would be replaced by calls to global error reporting function, which user could override. Something in lines of:
```
// turbojpeg.h

typedef void (*tjErrorReportingFn)(const char* format, ...);
DLLEXPORT void tjSetGlobalErrorReportingFn(tjErrorReportingFn fn);
```
```
// turbojpeg.c

static void defaultGlobalErrorReportingFn(const char* format, ...)
{
  va_list args;
  va_start(args, format);
  vsnprintf(errStr, JMSG_LENGTH_MAX, format, args);
  va_end(args);
}

static tjErrorReportingFn globalErrorReportingFn = &defaultGlobalErrorReportingFn;

DLLEXPORT void tjSetGlobalErrorReportingFn(tjErrorReportingFn fn)
{
  globalErrorReportingFn = fn;
}

// and replacing every "snprintf(errStr, JMSG_LENGTH_MAX, " (and also without the space at the end) with "globalErrorReportingFn("
```

That way any thread-aware program can take over the burden of handling global error strings and properly assign them to the correct threads (getting the calling thread id in the supplied error reporting function, storing the value somewhere), or, if no global error reporting is required, simply disable that completely by supplying an empty function like this:
```
void myTjErrFn(const char* format, ...)
{
}

int main(int argc, char **argv)
{
  tjSetGlobalErrorReportingFn(&myTjErrFn);
  ...
```

But if you can think of any better solution which allows bypassing the thread clashes, I am all for it.

**Comment by dcommander** [comment]
Can you provide an example of a test that fails with TSan?

I don’t understand why your proposal above is an improvement. You are simply replacing one unguarded global variable with another.  What is to prevent one thread from calling `tjSetGlobalErrorReportingFn()` while another thread encounters an error? The worst case at the moment (which I posit would never occur unless the calling program was buggy) is a junk error string. The worst case with your design is crashing the program.

The calling program cannot expect thread safety if it passes an invalid handle to an API function, because the handle points to a structure that contains the instance-specific information necessary to guarantee thread safety. So it is unreasonable to expect the TurboJPEG API to handle that case differently than it currently does. So what is left? Primarily your argument is that the TurboJPEG API should handle invalid arguments in a thread-safe manner. It mostly does, except in the case of functions that do not accept a handle. I am not interested in arguing hypotheticals, so I think the best way to approach this is to demonstrate reasonable workflows that cause errors with TSan, and I will address those errors. Perhaps that will require introducing new API function signatures so that all functions accept a handle. I can also look at other imaging APIs to see how they handle this issue, but first I want a concrete test that demonstrates the failures.

**Comment by pac-work** [comment]
As for the example of TSan failure, sure, see the code copy+paste at the bottom.

My proposal above is an improvement, because, as I tried to suggest by the code snippet, you call `tjSetGlobalErrorReportingFn` only once at the beginning of program using `libjpeg-turbo`, for example as a first line in the `main` function, before you start any threads. That way, you have set global error handling function in a thread safe way and within that function, you may do whatever you need to get proper full thread safety for the platform you are on.

```C++
// File name: libjpeg-turbo-is-unsafe.cpp
// libjpeg-turbo compiled with:
//   cmake -DCMAKE_INSTALL_PREFIX="../libjpeg-turbo.install" -DCMAKE_C_COMPILER="/usr/bin/clang" -DCMAKE_C_FLAGS="-fsanitize=thread -g" ../libjpeg-turbo
//   make -j8
//   make install
// This source code compiled with:
//   clang++ -Wall -fsanitize=thread -I./libjpeg-turbo.install/include/ -L./libjpeg-turbo.install/lib -lpthread -lturbojpeg -Wl,-rpath=./libjpeg-turbo.install/lib libjpeg-turbo-is-unsafe.cpp

#include <thread>
#include <vector>

#include <turbojpeg.h>

void threadFn()
{
  // Just 2 of many possible examples, uncomment one by one:

  // Example of wrong parameters causing UB, although API documentation says: @return the maximum size of the buffer (in bytes) required to hold the image, or -1 if the arguments are out of bounds
  // This could be theoretically prevented by explicitly checking all the parameters of all the calls to the library before doing the calls, virtually copy+pasting the source code of the library to the application. That does not really feel elegant.
  /* tjBufSize(0, 0, 0); */

  // But not all the problems can be detected upfront... Even if you check for the file existance, it can be deleted in between the check and the call.
  /* int width, height, pixelFormat;
  pixelFormat = TJPF_UNKNOWN;
  tjLoadImage("/NON_EXISTING_FILE", &width, 1, &height, &pixelFormat, 0); */ 
}

int main()
{
  const size_t numOfThreads = 2;
  std::vector<std::thread> threads;
  threads.reserve(numOfThreads);
  for (size_t i = 0; i < numOfThreads; ++i)
    threads.push_back(std::thread(&threadFn));
  for (size_t i = 0; i < numOfThreads; ++i)
    threads[i].join();
  return 0;
}

/*--------------------------------------------------------------------------------------------------------------------*/

/* Output for the first example:
==================
WARNING: ThreadSanitizer: data race (pid=18381)
  Write of size 8 at 0x7fa04af2a200 by thread T2:
    #0 memcpy <null> (a.out+0x4301cd)
    #1 tjBufSize /home/pac/Tmp/libjpeg-turbo/turbojpeg.c:498:5 (libturbojpeg.so.0+0x862ed)
    #2 threadFn() <null> (a.out+0x4b0b5d)
    #3 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) <null> (a.out+0x4b2dbd)
    #4 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) <null> (a.out+0x4b2cf0)
    #5 decltype(std::__invoke(_S_declval<0ul>())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) <null> (a.out+0x4b2c98)
    #6 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() <null> (a.out+0x4b2c38)
    #7 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() <null> (a.out+0x4b2a3c)
    #8 <null> <null> (libstdc++.so.6+0xbbb2e)

  Previous write of size 8 at 0x7fa04af2a200 by thread T1:
    #0 memcpy <null> (a.out+0x4301cd)
    #1 tjBufSize /home/pac/Tmp/libjpeg-turbo/turbojpeg.c:498:5 (libturbojpeg.so.0+0x862ed)
    #2 threadFn() <null> (a.out+0x4b0b5d)
    #3 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) <null> (a.out+0x4b2dbd)
    #4 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) <null> (a.out+0x4b2cf0)
    #5 decltype(std::__invoke(_S_declval<0ul>())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) <null> (a.out+0x4b2c98)
    #6 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() <null> (a.out+0x4b2c38)
    #7 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() <null> (a.out+0x4b2a3c)
    #8 <null> <null> (libstdc++.so.6+0xbbb2e)

  Location is global 'errStr' of size 200 at 0x7fa04af2a200 (libturbojpeg.so.0+0x000000101200)

  Thread T2 (tid=18384, running) created by main thread at:
    #0 pthread_create <null> (a.out+0x426845)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xbbdb4)
    #2 main <null> (a.out+0x4b0c0c)

  Thread T1 (tid=18383, finished) created by main thread at:
    #0 pthread_create <null> (a.out+0x426845)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xbbdb4)
    #2 main <null> (a.out+0x4b0c0c)

SUMMARY: ThreadSanitizer: data race (/home/pac/Tmp/a.out+0x4301cd) in memcpy
==================
ThreadSanitizer: reported 1 warnings
*/

/*--------------------------------------------------------------------------------------------------------------------*/

/* Output for the second example:
==================
WARNING: ThreadSanitizer: data race (pid=18390)
  Write of size 8 at 0x7f3f7b2ad200 by thread T2:
    #0 vsnprintf <null> (a.out+0x43617e)
    #1 snprintf <null> (a.out+0x437ac0)
    #2 tjLoadImage /home/pac/Tmp/libjpeg-turbo/turbojpeg.c:2015:5 (libturbojpeg.so.0+0x908d1)
    #3 threadFn() <null> (a.out+0x4b0b84)
    #4 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) <null> (a.out+0x4b2ddd)
    #5 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) <null> (a.out+0x4b2d10)
    #6 decltype(std::__invoke(_S_declval<0ul>())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) <null> (a.out+0x4b2cb8)
    #7 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() <null> (a.out+0x4b2c58)
    #8 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() <null> (a.out+0x4b2a5c)
    #9 <null> <null> (libstdc++.so.6+0xbbb2e)

  Previous write of size 8 at 0x7f3f7b2ad200 by thread T1:
    #0 vsnprintf <null> (a.out+0x43617e)
    #1 snprintf <null> (a.out+0x437ac0)
    #2 tjLoadImage /home/pac/Tmp/libjpeg-turbo/turbojpeg.c:2015:5 (libturbojpeg.so.0+0x908d1)
    #3 threadFn() <null> (a.out+0x4b0b84)
    #4 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) <null> (a.out+0x4b2ddd)
    #5 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) <null> (a.out+0x4b2d10)
    #6 decltype(std::__invoke(_S_declval<0ul>())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) <null> (a.out+0x4b2cb8)
    #7 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() <null> (a.out+0x4b2c58)
    #8 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() <null> (a.out+0x4b2a5c)
    #9 <null> <null> (libstdc++.so.6+0xbbb2e)

  Location is global 'errStr' of size 200 at 0x7f3f7b2ad200 (libturbojpeg.so.0+0x000000101200)

  Thread T2 (tid=18393, running) created by main thread at:
    #0 pthread_create <null> (a.out+0x426845)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xbbdb4)
    #2 main <null> (a.out+0x4b0c2c)

  Thread T1 (tid=18392, finished) created by main thread at:
    #0 pthread_create <null> (a.out+0x426845)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xbbdb4)
    #2 main <null> (a.out+0x4b0c2c)

SUMMARY: ThreadSanitizer: data race (/home/pac/Tmp/a.out+0x43617e) in __interceptor_vsnprintf
==================
ThreadSanitizer: reported 1 warnings
*/
```

**Comment by dcommander** [comment]
But those are artificial examples. An equally artificial example would be to call your proposed `tjSetGlobalErrorReportingFn()` function from multiple threads. My point is: thread safety currently relies on specific behaviors on the part of the calling program, and your proposed addition does not change that fact. It just changes the specific behaviors required. I will examine how other imaging APIs deal with this problem.

**Comment by pac-work** [comment]
I do not really understand, what is so artificial about something as known as filesystem race conditions (https://en.wikipedia.org/wiki/Race_condition#File_systems), but obviously, opinions may differ...

You are saying, that "thread safety currently relies on specific behaviors on the part of the calling program, and your proposed addition does not change that fact". I agree with that on the theoretical level, there is however one huge difference. Because if you want to achieve thread safety with my potential additions, you do 1 thing in the whole program to be able to use the library safely. You call 1 (potentially very well documented) function and that is it, no need to worry about anything else, no need to read implementation of every function you want to call and copy+paste all the code which does input checking, trying to avoid cases when things will fail (which is sometimes not possible at all, as I have shown with the FS example). That is why I chose the `tjBufSize` as the second example function. Looks very simple from outside, documentation clearly states how it should behave and nobody would expect this to cause UB, but it does. You are saying that every user of the library who wants thread safety needs to read complete code of the library to catch every corner case there might be to prevent any errors (what about errors in input data you are trying to decode? would they cause same problem? how should I know as a library user, when I have to handle any other potential error state?). I am saying that you call 1 function and you get 100% thread safety, no questions, no exceptions, no corner cases, no room for potential mistakes. As somebody who would like to use the library, for me, this is obvious choice. But again, as I see, opinions may differ.

The changes I proposed were trying to be as minimal towards the API as possible, affecting 0 existing applications, without the need for any backward compatibility maintenance. But if you find preferable to change the whole API to strictly always take handle and _never ever_ write into internal shared state (better to remove that one just to be sure), I cannot but agree that that would be a superior solution, only it would take much more work.

**Comment by dcommander** [comment]
"Because *San produces errors" is a useful argument, but it's not an ironclad argument.  Sanitizers are automated tools that are meant to highlight potential problem areas, but ultimately a human must decide how serious those highlighted problems are.  I agree with you that these problems exist.  I just disagree that they are serious.  Like I said, the worst case is probably garbling the error string.  (I should also note that I personally use the TurboJPEG API to perform multithreaded encoding and decoding in my own software, so I'm not just speaking in the abstract.)  A much more compelling argument would be to demonstrate why these issues represent a significant problem for developers writing real applications.  In other words, show me how to make TSan light up by doing (a) something that a real developer might actually do and (b) something that a real developer couldn't do any other way.

My point is:
It's pretty easy, when writing a multithreaded program that uses the TurboJPEG API, to ensure that functions like `tjLoadImage() and `tjBufSize()` are never called in parallel.  Doing that is much less onerous for a programmer than designing their own thread-safe error handling callback function and calling a separate API function to institute that error handler.  One of the reasons why the TurboJPEG API exists is to provide a more straightforward interface than the libjpeg API.  As soon as we start talking about handlers and callback functions, that undercuts the straightforwardness argument of the TurboJPEG API.

What I said above is that I need to investigate this further.  Arguing with me is distracting from that investigation.  I understand and have fully absorbed everything you said, but I disagree with your solution, for reasons stated above.  I would prefer introducing a pthreads dependency before introducing a new API function, but I need to do further research and see if there might be a better option than either.  I have other TurboJPEG API improvements in mind for the next major release, so I need to think about this set of problems in that context.  It's not something that will be a quick fix.  I may opt to, for instance, change the behavior of most of the API functions so that they return multiple error codes based on the specifics of the error (-1 for an instance error that requires calling `tjGetErrorStr2()`, -2 for an invalid argument, etc.)  Like I said, I need to look at how other imaging APIs solve this problem.  I'm planning to make some broad changes to how the TurboJPEG API operates (moving more in the direction of state setting/getting functions rather than using flags and parameters to control the operation), so that provides an opportunity to revisit how errors are handled in general.

**Comment by pac-work** [comment]
Thank you for your answer. Well, we obviously have very different view on how important is to have software which is without undefined behavior, stable, and stays that way. The view similar to “I know it is broken, against the standard and even sanitization tools report that, but I do not find it serious, because it kind of works with my current compiler, standard library, architecture, compiler options, use case, etc… and problems caused by it will not happen often and seems to not be that serious for me at this moment...”, is hardly understandable for me... You are telling me that you want more “real world” examples. I have shown you how even state external to application (filesystem state, maybe data you are decoding,...) can cause undefined behavior in the application and that is not enough for you. Any serious security expert would just start crying now. But that is, as it seems, really matter of opinions and previous experience.

I agree that there are workarounds, of course. And I am using one of them (posting all the JPEG compression related work to one thread only, which is enough for me now), but all of them will essentially work as a global lock. Whether these workarounds are simpler than one function call and an error handler is from my point of view again very questionable (especially because in one case, you get full performance of your HW, in the other one not).

Well then, I will not waste more of your time now, as you suggested, and will be looking forward to some API changes which help to make things right. Unfortunately, at this point, I really do not have anything to contribute.

Anyway, thanks for the work done on the TurboJPEG, otherwise I find it very good.

**Comment by dcommander** [comment]
Give it a rest, OK? I said I want to fix this. I just disagree with your proposed fix. There are a lot of constraints I have to deal with that you are unaware of. Go maintain the world’s most popular JPEG library for 10 years and tell me how easy it is for you. I have ultimately fixed every major issue that anyone has discovered in libjpeg-turbo, but when millions of people use your code on a daily basis, a great deal of care is required when implementing fixes and new features. I will fix this the right way, but it will take some time and will not land until 2.1. Please be patient. Spewing your vitriol at me is certainly not going to make this happen any faster.

**Comment by dcommander** [comment]
Well, I now know way more about cross-platform thread-local storage than I ever wanted to know.  The good news is that this was fixed non-disruptively on all supported platforms except 32-bit macOS (which doesn't support TLS), and the fix will be in 2.0.5.  The bad news is that testing the fix on all of those platforms exhausted 2 1/2 months worth of project funding.

Please do me a favor and test the fix.  Thanks.

**Comment by dcommander** [comment]
NOTE: The fix has been applied to the iOS 64-bit (ARMv8) fork by increasing the minimum supported version of iOS from 7.0 to 8.0.  It could similarly be applied to the macOS 32-bit fork by increasing the minimum supported version of macOS from 10.5 to 10.7 and building with Xcode instead of GCC, but I chose not to do so because the macOS 32-bit fork is intended to support obsolete platforms.  So basically the fix is now present in all official binaries that support non-obsolete platforms, and only 32-bit Apple binaries lack the fix.  32-bit iOS support has already been removed from libjpeg-turbo 2.1.x, since there is no way to deploy 32-bit apps to iOS devices anymore (the iOS App Store now requires 64-bit apps.)  32-bit macOS support will eventually be removed as well, but not in libjpeg-turbo 2.1.x.

**Comment by pac-work** [comment]
>     3. The worst case if the behavior is encountered truly is limited to a garbled error string (I verified this in the process of painstakingly testing the fix.)
> 
> 
> However, what's done is done. Please do me a favor and test the fix. Thanks.

Hello, sorry for late answer, was on holiday.

I have tested version 2.0.5 with my application on my architecture and compiler (amd64 + Clang 7.0.1) and everything behaves as it should. Thank you very much for the effort you put into fixing this!

I really think it was worth the time and effort, because the fact that it now did not cause worse problems than garbled error string (in the tested context) does not mean that it will be like that in the future (with better optimizations done by compilers, with LTOs enabled, ...). Plus, UB is always a bad thing even from the process point of view (company quality controls, etc.).

Again, thanks a lot! This non-intrusive fix is definitely better than adding something to the API.

**Comment by njk42** [comment]
Hello,  I stumbled upon this post when searching for multi-threading help related to using the TurboJPEG C API (2.0.6) for DEcompression.  I hope you can help me determine if what I'm doing is simply incorrect usage of the TurboJPEG library (with expected results when used this way) or if there is some underlying bug.  I don't see anything about thread safety of the API anywhere in the documentation other than for the tjGetErrorStr2() function. 

My current approach does the following (in pseudo-code):

Main thread:
   handle = tjInitDecompress();
   ...then spawn two Worker threads...

Worker1:
    while( ) {
      tjDecompressHeader3(handle, )
      tjDecompress2(handle, )
    }

Worker2:
    while () {
      tjDecompressHeader3(handle, )
      tjDecompress2(handle, )
    }

Main thread:
   ... wait for worker threads to finish and join...
   tjDestroy();


I am getting perfect functionality with only one Worker thread, but random crashes when two Worker threads are active.  

I want to take advantage of multiple cores in my processor for parallel decompression (the JPEGs being decompressed by the two workers are completely unrelated), so guarding the activity with a mutex in the Worker threads defeats the purpose.  Do I need to use the more complicated lib-jpeg to accomplish this?  Does tjDecompress2() depend on some global state that is set during the prior call to tjDecompressHeader3() which would cause issues if these two functions are not called atomically?

Thanks for your response!

**Comment by dcommander** [comment]
@njk42 You have to use a different `tjhandle` instance for each thread.

**Comment by njk42** [comment]
OK, thanks for the amazingly fast response!

**Comment by Seneral** [comment]
This is a bit of a problem if you're using turbojpeg in multiple threads you can't easily clean up (e.g. a threadPool) but also don't want to init and destroy a tj instance every time you use it.
Here's a short snippet how I used unique_ptr to prevent memory leaks (not super important, but loud and annoying in Leak Sanitizer):

```C++
#include <memory> // unique_ptr
#include <turbojpeg.h>

struct tjDeleter
{
	void operator()(tjhandle handle) const
	{
		tjDestroy(handle);
	}
};

using tjInstance = std::unique_ptr<void, tjDeleter>; // for tjHandle which is a void*

thread_local tjInstance tjCompressInstance = nullptr;

// Later before use:
{
	// Init TJ
	if (!tjCompressInstance && (tjCompressInstance = tjInstance(tjInitCompress())) == NULL)
	{
		return -1;
	}
        // Use TJ thread-safely
}
```

**Comment by dcommander** [comment]
@Seneral That isn't really the same issue. The expectation is that each thread must have a separate TurboJPEG instance in order to achieve thread safety, so this issue was about a problem in the library's design that caused it to deviate from that expectation.

**Comment by Seneral** [comment]
> @Seneral That isn't really the same issue. The expectation is that each thread must have a separate TurboJPEG instance in order to achieve thread safety, so this issue was about a problem in the library's design that caused it to deviate from that expectation.

I am aware, but this was a thread that came up during research, and that's where I think such a snippet makes the most sense. It was in reply to your latest comment, should've clarified:

> @njk42 You have to use a different `tjhandle` instance for each thread.

I just arrived here because I had hoped there was a better way to use TJ multithreaded multithreaded by now, or at least not need a preallocation anymore, since the combination of that makes it a bit inconvenient to use at times. I've always used thread_local, but that left a memory leak, which I've finally addressed with the snippet.

**Comment by dcommander** [comment]
> I just arrived here because I had hoped there was a better way to use TJ multithreaded multithreaded by now, or at least not need a preallocation anymore, since the combination of that makes it a bit inconvenient to use at times. I've always used thread_local, but that left a memory leak, which I've finally addressed with the snippet.

"Better" is subjective and depends a lot on your specific use case.  This issue was the first time that anyone complained about the TurboJPEG API in a multithreaded environment.  The API's design is consistent with that of other image codecs, most of which can also only claim thread safety on a per-instance basis.  Furthermore, the API's design is constrained by the design of the underlying libjpeg API, which can also only claim thread safety on a per-instance basis.  Your code still creates a different `tjhandle` instance for each thread.  It just does that semi-automatically using `thread_local` and `std::unique_ptr` (a C++11 feature.)  Applications are free to do that as they see fit, but obviously I cannot take advantage of a C++11 feature in a C library.  I frankly find it more readable when object-oriented code wraps a C API instance at the class level and uses a separate class instance for each thread, rather than relying upon constructs like `thread_local` and `unique_ptr`.  To each their own.  I am certainly open to suggestions for how to improve the TurboJPEG API, and I would encourage you to make those suggestions under https://github.com/libjpeg-turbo/libjpeg-turbo/issues/646.  (I would love for you to elaborate on your comment about preallocation, for instance.)

libjpeg-turbo, as a project, has an extremely limited budget, so almost all major enhancements to it in its 15-year history have been the result of funded development.  The TurboJPEG API actually predates this project.  Around the turn of the millennium, I was working at Landmark Graphics (an oil & gas exploration and production software company) on proprietary solutions for remotely displaying their seismic visualization applications, which often used datasets that were too big to transfer over the network.  At the time, I was using the Intel JPEG Library (IJL) as a high-speed image transport for OpenGL-rendered frames.  When Intel dropped the IJL in favor of the lower-level Integrated Performance Primitives (IPP), I wrote a higher-level wrapper around both the IPP and libjpeg that exposed just the features I needed, and that wrapper was open sourced along with VirtualGL and TurboVNC in 2004.  During the following four years, in which VirtualGL and TurboVNC were productized by Sun Microsystems, the wrapper was renamed TurboJPEG and became a cross-platform abstraction layer for various lower-level JPEG codecs (IPP, Sun mediaLib, Apple Quicktime, Pegasus Imaging PICTools, and libjpeg.)  After my entire team at Sun was laid off in early 2009, Cendio AB hired me to add 64-bit SSE2 extensions and Huffman optimizations to an in-tree version of libjpeg/SIMD they were using in TigerVNC (before it was even called TigerVNC.)  As that in-tree codec improved, other projects wanted to use it, so I was encouraged to split it into a standalone project.  Initially, libjpeg-turbo was just a "you-might-find-this-useful" library that installed under **/opt**, so applications that needed high-speed JPEG could load it instead of the system's libjpeg implementation.  I included the TurboJPEG API in case people might find that useful as well, and also to facilitate "eating my own dog food" by building VirtualGL and TurboVNC against libjpeg-turbo.  At the time, I certainly couldn't have dreamed that libjpeg-turbo would become a de facto industry standard, much less an official ISO/ITU-T reference implementation.

That's a roundabout way of getting to the point that the TurboJPEG API was originally just a quick & dirty way to access the fastest underlying libjpeg-turbo features.  It has gained momentum in recent years largely because the libjpeg API is showing its age, so I have been encouraged to expose more and more underlying libjpeg API features in TurboJPEG (including some that aren't actually "turbo" per se.)  However, the API has received no direct funding in the past 10 years.  Thus, all of the enhancements made to it since TurboJPEG 1.4.x, including the extensive TurboJPEG 3 overhaul, have been made incrementally using General Fund money (which derives from both corporate and individual donations.)  The API evolved more than it was designed, so it picked up a lot of oversights.  TurboJPEG 3 was meant to address most of the oversights that users had reported over the years, but I have seen very little adoption of it (even among people who already use the TurboJPEG API.)  I have also received very little feedback, and in the absence of feedback, I have to make executive decisions.  TurboJPEG 4 will be another opportunity to address remaining oversights, including perhaps multithreaded usability, but I have no idea when or if enough funding will accrue to make that happen.  (At the moment, our project is running at a deficit, which is often the case after a major beta release.)

**Comment by Seneral** [comment]
(Sorry if this is all jumbled, it's late and I really should be sleeping)
I can certainly sympathise with projects growing to a scale it may initially never been designed for, and I have immense respect for sticking to it and making it as good as it is rn. Though tbh I'm not a power-user of libjpeg-turbo, I just saw a typical C API and didn't think much of it at the time. I just happened to write a wrapper for it to fit my expectations - more on that in a bit.
In general, there's no API to do it all anyway. I don't know much about the internals, and I don't think this interface is wrong or anything, and really I'm in no position to judge. There are many programming styles, and there's no way you'd make everyone perfectly happy, so I do think the plain C interface is best. It's just that the way I use the library happens to involve multiple threads, and multiple places I'd like to call it, and no guarantees about thread lifetime, and I can't be bothered to write some "manager of images" construct that's thread-safe and does it all or something. See me as a user that really just wants a quick and simple JPEG library - but cares just enough about performance to go beyond the simpler header-only implementations (though tbh I never really tested them, the decision to go with libjpeg-turbo was made a few years back).

If it helps a bit, I guess I can serve as a tiny sample of the userbase and my requirements and expectations, as simple as they may be. My project is still WIP, and JPEG is not a major, central part of it - in short, 8bit grayscale encoding on custom cameras (Raspberry Pis), sending over a custom bandwidth-limited protocol to a PC, for visualisation purposes only, about 2-80KB per image. The encoding side will eventually be replaced by a hardware encoder that's already there, so really, the PC side is fine with a "decode this with whatever you see fit" simple API that 1.x provides.

My thoughts on the API, especially tjhandle:
If the internals need structures of some kind to work (that's what I just blindly assume the tjhandle represents), then exposing that for optimal performance for projects that need it makes sense. If a "simpler" API were to instead allocate memory on-demand and hide that complexity, I don't think I'd be any happier. If it's already allocating new temporary structures anyway, then maybe my only critique is that it's not clear what the handle is for in the first place.
Dumbing it down: Why do I need it at all? Can't I just initialise it and then destroy it immediately every time I use it? Would that be a huge loss in performance? - again, I obviously didn't care enough to try to dig deeper to really find these answers myself.

For me, I really just wanted a simple functional API (I don't really do OOP much), so the thread_local storage and set-and-forget construction/deconstruction of a TJ instance was the least complicated solution. Only now when I used leak sanitizer I found it to be leaking, so I fixed it, even if it required some weird C++11 trickery that personally, I am not too acquainted with either - hence me sharing it on a whim. Though I did switch to a less-obviously-modern-C++ solution, see my edit. Still thread_local, that's just too convenient.

So really I made that C++11 wrapper to simplify the interface for my purposes. I don't really expect this project, being a C library, to cater to C++ at all, or for it to sacrifice it's core value of speed, or compatibility, for a simpler API. I don't think that's the job of libjpeg-turbo. I wouldn't however mind an officially sanctioned snippet like the above for simpler, functional, thread-safe use in C++. Link it in the readme, put in an example, or let SEO do it's work and let my comment serve that use (which is enough, hence I put it here where I expected it to be found if people search for it). Really doesn't matter too much.

If I had one complaint it's distribution. As an open source project using this library without a proper distribution myself, I don't want my users to have to manually install libjpeg-turbo. But I also didn't find an easy solution to compile it locally via scripts (which I do for other larger dependencies that I can't just integrate manually), since it requires it's own tools to compile optimally (NASM/yasm). I have no good solution for this right now - currently I am literally using 7zip in a script to unpack the self-extracting exe for windows, and ask users to install the package using their package manager for linux. I would prefer my scripts to be able to download the library binaries directly (for static linking) without user intervention, without installing anything on the users system. I would prefer to be able to compile it, but I would not go through that hassle for SSE optimisations if I'm being honest.
I will likely reevaluate before release what I really want out of a JPEG decompressing solution, and if the (probable, honestly untested) extra performance is worth the distribution hassle, or if I should just switch to a header-only solution. That is not something I intended to do now though.
Again, I see libjpeg-turbo as a performance-first kinda library, and not the general-JPEG-library for C++ that I use it as. I don't think I'm a typical user, and I don't think you should cater to my use case too much. If API is a hindrance for anybody, then distribution would be much more so, and really, if you want the performance, you probably don't care about either of them too much. I just need to reevaluate where I lie on that balance eventually - and honestly, this post made me think about it much more than I should, but here you go. Hope it helps. Thanks for the library!

**Comment by dcommander** [comment]
> My thoughts on the API, especially tjhandle: If the internals need structures of some kind to work (that's what I just blindly assume the tjhandle represents), then exposing that for optimal performance for projects that need it makes sense. If a "simpler" API were to instead allocate memory on-demand and hide that complexity, I don't think I'd be any happier. If it's already allocating new temporary structures anyway, then maybe my only critique is that it's not clear what the handle is for in the first place. Dumbing it down: Why do I need it at all? Can't I just initialise it and then destroy it immediately every time I use it? Would that be a huge loss in performance? - again, I obviously didn't care enough to try to dig deeper to really find these answers myself.

Yes, `tjhandle` is a pointer to an opaque struct, and among other things, that struct stores the underlying libjpeg API structs.  Even if written in a non-object-oriented language, most APIs have the concept of objects and instances.  Each `tjhandle` represents an instance of the TurboJPEG API object, as well as an instance of the underlying libjpeg API object.  Think of it like a class instance in C++.  You can't expect a C++ API to manage class instances "behind the scenes" on behalf of the calling application.  The basic contract of OOP is that the calling application manages object instances, and methods are called against a specific instance.  So you can think of the TurboJPEG function calls that accept a `tjhandle` argument as methods that are called against a specific TurboJPEG API instance.  There is no practical way for an API to manage instances "behind the scenes" unless it has some facility to bind those instances to another type of object, such as a thread.  For instance, OpenGL has the concept of "contexts", which are essentially instances, but you have to explicitly make them current within a thread.  The result is that you don't have to specify the instance (context) handle with every OpenGL function call, but you do have to make more function calls in order to create and bind the instance.  That design made sense for OpenGL because contexts are platform-specific, but with respect to TurboJPEG, making a similar design decision would make one use case easier at the expense of making many other use cases impossible.  Furthermore, it would introduce a different calling convention to that of the underlying libjpeg API.  Per above, TurboJPEG originated as a wrapper around a handful of lower-level APIs, and all of those lower-level APIs had/have similar object/instance constructs.

> If I had one complaint it's distribution. As an open source project using this library without a proper distribution myself, I don't want my users to have to manually install libjpeg-turbo. But I also didn't find an easy solution to compile it locally via scripts (which I do for other larger dependencies that I can't just integrate manually), since it requires it's own tools to compile optimally (NASM/yasm). I have no good solution for this right now - currently I am literally using 7zip in a script to unpack the self-extracting exe for windows, and ask users to install the package using their package manager for linux. I would prefer my scripts to be able to download the library binaries directly (for static linking) without user intervention, without installing anything on the users system. I would prefer to be able to compile it, but I would not go through that hassle for SSE optimisations if I'm being honest. I will likely reevaluate before release what I really want out of a JPEG decompressing solution, and if the (probable, honestly untested) extra performance is worth the distribution hassle, or if I should just switch to a header-only solution. That is not something I intended to do now though. Again, I see libjpeg-turbo as a performance-first kinda library, and not the general-JPEG-library for C++ that I use it as. I don't think I'm a typical user, and I don't think you should cater to my use case too much. If API is a hindrance for anybody, then distribution would be much more so, and really, if you want the performance, you probably don't care about either of them too much. I just need to reevaluate where I lie on that balance eventually - and honestly, this post made me think about it much more than I should, but here you go. Hope it helps. Thanks for the library!

All open source projects require some external build tools.  The biggest one is usually the compiler.  The libjpeg-turbo Windows/x64 build also requires CMake and NASM or Yasm, both of which are available in standalone Windows installers that are a lot easier to install than any Windows compiler.  So I'm not sure what you expect.  Building with just a `configure` script isn't a Windows thing and never has been.  Also, there are independent downstream packagers, such as vcpkg, that provide libjpeg-turbo.  We could distribute a .zip file with the pre-built libraries, but that wouldn't save you any steps vs. using 7z to extract the installers.  Furthermore, you can always run the installers in silent mode.

Also, libjpeg-turbo is an ISO/ITU-T JPEG-1 reference implementation, and even Windows itself uses libjpeg-turbo behind the scenes.  Because of its presence in popular web browsers and operating systems, libjpeg-turbo has billions of active daily users.  I'm not sure how much more general-purpose you can get than that.  My historical notes above were meant to highlight the fact that the TurboJPEG API, and libjpeg-turbo itself, were originally much more fit-for-purpose rather than general-purpose.  However, that all changed as people stopped using libjpeg in favor of libjpeg-turbo.  (That occurred primarily because the current maintainer of the IJG focused on novel and non-conformant JPEG format extensions that allow the DCT algorithm to do things that it was never designed to do, rather than focusing on the usability and performance and modernization that users actually wanted.  Tom Lane, the creator of libjpeg and the original IJG maintainer, declared that libjpeg-turbo dovetails with his original goals for libjpeg.)

**Comment by Seneral** [comment]
Fair enough, what you say makes sense, I just wanted to state my experience since it seemed like you wouldn't mind some feedback, even if it's not about the later API version.
Ofc I've worked with other libraries that use contexts that are not thread-safe, but that's usually UI, libusb, and other things that would typically have their own thread anyway for complex logic handling, and sometimes dedicated mechanisms to enable some form of thread-safe communication that makes sense for their specific uses.
Without knowing the internals of libjpeg-turbo, specifically that libjpeg itself requires such a context, I would not have expected such an interface from an image compression library, rather I'd expected it to be functional in nature - e.g. one call does not share state with another and thus it'd be inherently thread-safe. Even now I still don't know what the "context" for a image decompression library would even entail, since the functions themselves should, in theory, not need to mutate any global state. As I said, my best guess was preallocated memory to speed up calculations, now I know it also has to do with underlying libjpeg, which is all good and well, but I couldn't have known that. And it doesn't change the fact that I personally will use the code in my project in a functional manner, hence the wrapper. It's not like the libusbs or the graphics/UI libraries of my project that have a dedicated thread, they're "decompress whereever, whenever is needed". Again, I don't expect that to change (I don't think it's something you can or should fix), but it is a slight mismatch of expectation vs reality nonetheless, from a user not too versed with typical interfaces of C libraries. Creating a wrapper is easy enough, but just nuanced enough I felt like sharing it on a whim (since I obviously didn't get it "right" on the first try, with an unimportant memory leak).

Regarding distribution, providing a .zip with all build artifacts would make things easier. On windows I currently rely on 7zip to be installed, and if it's not, I again just have to ask the user to either install it or manually run the .exe (I am strictly against modifying the users system outside my projects folders with scripts). On linux, I can't do anything in that regard, though I already ask users to install other packets so asking them to install one more is not a huge deal. But, considering libjpeg-turbo is the only one of those dependencies that I am actually statically linking into my project binary itself (the rest being build tools only), I still would prefer just a zip.
But I'm also not super versed with build systems, especially on windows, and I don't want my users to need to be either. So asking them to fetch all dependencies from multiple different sources, even if one of them is vcpkg, is not something I want. Having a downloadable zip in the releases on github would be good to have as an option, IMO.
This won't be a problem for most big users of yours I imagine, if you're a project shipping with an installer anyway, adding libjpeg-turbo to be installed on the users system is no big deal, even a bonus. I'm just coming from a different viewpoint, as I explained, and there's no need to indulge me if you don't like it. But it is true that for my use, libjpeg-turbo is by far my single most annoying dependency to deal with, at least after I got rid of wxWidgets and OpenCV (which I at least was able to compile locally without installing other things). And for a project like mine, that might be more of a deciding factor than performance. That's yet to be determined ofc, really not that important right now.
